File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.py
def solution(numbers):
    answer = [-1 for i in range(len(numbers))]
    
    stack = []
    
    for idx, target in enumerate(numbers):
        while stack and target > numbers[stack[-1]]:
            answer[stack.pop()] = target
        
        stack.append(idx)
        
    return answer

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.Stack;

class Solution {
    public int[] solution(int[] numbers) {
        int n = numbers.length;
        int[] answer = new int[n];
        for (int i=0; i<n; i++) {
            answer[i] = -1;
        }
        
        Stack<Integer> stack = new Stack();
        for (int i=0; i<n; i++) {
            int target = numbers[i];
            
            while (!stack.empty() && target > numbers[stack.peek()])
                answer[stack.pop()] = target;
            
            stack.push(i);
        }
        
        return answer;
    }
}